lib: os: cbprintf: Add static packaging

nordic-krch opened the pull request:
Alternative take to #31102, based on #32734.

`Statically` means that a package (serialized string with arguments) is created as assignment to a buffer. Opposed to runtime package creation with cbprintf_package() where string is analyzed and based on format specifiers a package is created (any transient strings are copied into the package buffer).

In principle, `CBPRINTF_STATIC_PACKAGE(buf, &len, "test %d", n)` resolves to simple assignments.

`CBPRINTF_MUST_RUNTIME_PACKAGE` is used to determine at compile time if static packaging can be applied.
`CBPRINTF_STATIC_PACKAGE(NULL, &len, "test %d", n)` - null buffer results in length calculation

Package on stack can be created with following code:
```
size_t len;
CBPRINTF_STATIC_PACKAGE(NULL, &len, "test %d", n);
uint8_t buf[len]; /* len is determined at compile time */
CBPRINTF_STATIC_PACKAGE(buf, &len, "test %d", n);
```

Note that static packaging is using `_Generic` C11 keyword with fallback to runtime packaging if not available and local suppressing of  `pointer-arith` warning (`(arg) + 0` generates warning if arg is `void *`).

nordic-krch committed:
lib: os: cbprintf: Add static packaging

Extend cbprintf with static packaging.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: cbprintf_package: Extend to test static packaging

Extend test to validate cbprintf static packaging on various
platforms.

Added more test configurations: complete, nano, long double.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

andyross commented:
I've been busy with other stuff and avoiding the printf drama.  And FWIW this code looks good to me.

But... is this solving the wrong problem?   As I understand it, the goal behind deferring formatting is mostly to get the output of the log subsystem serialized in a form that does not require formatting on the device at all, right?   How much surgery to our printf layer is really appropriate given that as a goal?

I was looking at dictionary logging at one point too, and my (somewhat vague) plan was something more like:

1. Put the format string in a separate non-loadable section and give it a unique value at runtime via linker magic
2. Serialize the arguments and the ID in a recoverable way at the point of logging
3. Hand that stuff to an offline python script to figure out
4. Profit

I mean, is this code even going to be run on deployed production device?  Or are we spending all this effort to come up with a scheme to allow devices to duplicate the production code in some kind of debug mode?

nordic-krch commented:
@andyross packaging it's not only about dictionary logging. Current logging scheme is storing all arguments into `void *` items. This limits logging data (no double, long long, etc.). Additionally, for `%s` only string address is used and user must call `log_strdup` explicitly to dump the string into dedicated pool. Another aspect is multi-domain logging: `log_strdup` makes it harder to pass log messages between domains or copy a message. Same goes for dictionary logging or logging to flash: we would like to have self contained package that can be stored and formatted later. This is the motivation for having packaging at all. Now, static packaging is an extension to speed things up. If there are no `%s` in the string then package can be created statically as simple memory assignments, no need to walk through the string. With `_Generic` you can detect if there are any `char *` arguments and if not you can simply dump all arguments (utilizing `_Generic` to promote `floats` to double) and create the package. For dictionary backends, message string can be put into the section which is removed from the build and processing happens as you wrote.

npitre commented:
On Thu, 4 Mar 2021, Andy Ross wrote:

> How much surgery to our printf layer is really appropriate given that as a goal?

At least with the latest proposal the printing side is completely 
unaffected. The idea now is to pre-package a va_list-like buffer that 
can be consumed later.

> I was looking at dictionary logging at one point too, and my (somewhat 
> vague) plan was something more like:
> 
> 1. Put the format string in a separate non-loadable section and give it a unique value at runtime via linker magic
> 2. Serialize the arguments and the ID in a recoverable way at the point of logging
> 3. Hand that stuff to an offline python script to figure out
> 4. Profit

This is almost the case now. The va_list alike chunk only contains 
pointers to strings that are known to live in read-only memory, or 
packed into the chunk otherwise. So one could retrieve those chunks and 
perform the printing offline while omitting the actual printing code 
from the device entirely.

jukkar commented:
```
size_t len;
CBPRINTF_STATIC_PACKAGE(NULL, &len, "test %d", n);
uint8_t buf[len]; /* len is determined at compile time */
CBPRINTF_STATIC_PACKAGE(buf, &len, "test %d", n);
```
I do not know enough of this but isn't compiler generating VLA in this case?

nordic-krch commented:
@jukkar added following code to hello world:
```
  size_t len;
  CBPRINTF_STATIC_PACKAGE(NULL, &len, "test %d\n", 1);
  uint8_t buf[len];
  CBPRINTF_STATIC_PACKAGE(buf, &len, "test %d\n", 1);
  cbpprintf(out, NULL, buf);
```
Got following assemby on arm cortex-m4, compiler is able to calculate size of the package:
```
LDR         R3, =0x0000405F           /* string address*/
STR         R3, [SP, #8]              /* store it at offset 8 */
MOVS        R3, #1                    /* set argument */
STR         R3, [SP, #12]             /* store it at offset 12 */
MOVS        R1, #0                    /* second argument of cbpprintf is in R1, set to null */
MOVS        R3, #12                   /* size of the package: 12 bytes*/
ADD         R2, SP, #4                /* third argument of cbpprintf is in R2, set to package address SP+4*/
LDR         R0, =out                  /* first argument of cbpprintf */
STRH.W      R3, [SP, #4]              /* Store package length in the first field  uint16_t*/
STRH.W      R1, [SP, #6]              /* set second half to 0 */
BL          cbpprintf                 /* call formatting of the package */

npitre reviewed (commented):
Didn't know about that _Generic thing. That's nice! Looks good overall.

You removed static qualifiers in the unit test file.
I'm kind of guessing why you did that, but mentioning the reason explicitly
in the commit log would be best.

I stopped using `VA_STACK_ALIGN_LONGLONG`, `VA_STACK_ALIGN_DOUBLE`, etc.
so to use only VA_STACK_ALIGN(type) directly as nothing needs special
exception for those types. Would be a good idea if you did the same to
avoid possible confusion... and reduce the number of code lines.

Finally, the argument length is now `final_length / sizeof(int)` and
it is stored in the first uint8_t. I'm not sure if your code matches.
Or I didn't recognize that part.

nordic-krch committed:
tests: unit: cbprintf: Extend to test static packaging

Extended test suite to test static packaging, too.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@npitre 
> You removed static qualifiers in the unit test file.

Extracted into separate commit.

> I stopped using VA_STACK_ALIGN_LONGLONG, VA_STACK_ALIGN_DOUBLE

Removed.

> Finally, the argument length is now final_length / sizeof(int) and
it is stored in the first uint8_t.

Code updated.

nordic-krch commented:
@pabigot could you take a look?

pabigot commented on include/sys/cbprintf_internal.h:
```diff
@@ -0,0 +1,270 @@
+ *
+ * @param ... String with variable list of arguments.
+ */
+#define Z_CBPRINTF_STATIC_PACKAGE(buf, _len, ... /* fmt, ... */) \
```
Do we have existing uses of reference parameters in macros?  It's not a natural idiom in C.

nordic-krch replied:
that was initially mimicking function call which was taking length by reference. I will remove it.

nordic-krch replied:
given https://github.com/zephyrproject-rtos/zephyr/pull/32863#discussion_r591510391 i'm not sure that i understand now. So since `_len` is input and output should it be passed by pointer?

pabigot replied:
I'm saying that people generally won't expect macro parameters to be both input and output, because you can't do that in functions.  The way the behavior is accomplished in functions is to accept a pointer.  To use a different idiom for in+out parameters in macros just because they're implemented as textual substitution creates an opportunity for confusion.

This is my opinion, I'm not blocking on it.

nordic-krch replied:
updated to have `inlen` and `outlen` parameters.

pabigot commented on include/sys/cbprintf.h:
```diff
@@ -21,6 +21,21 @@
+ * - Clang 3.0 https://releases.llvm.org/3.0/docs/ClangReleaseNotes.html
+ */
+#ifndef __C_GENERIC
+#if ((__STDC_VERSION__ >= 201112L) || \
```
It might be better to integrate this into the introduction of the API.

Also `__C_GENERIC` is not a legal identifier for anybody but the toolchain to use.  Even if somebody wants to say "but we *are* the system so we can do whatever we want" it'll confuse people.  `Z_C_GENERIC` maybe.

nordic-krch replied:
> It might be better to integrate this into the introduction of the API.

do you mean squashing commits?

pabigot commented on tests/unit/cbprintf/main.c:
```diff
@@ -220,22 +240,52 @@ static int rawprf(const char *format, ...)
+	rc = prf(sp_buf, _fmt, __VA_ARGS__); \
+} while (0)
+
+#define TEST_PRF(rc, _fmt, ...) \
```
I'm really not a fan of reference parameters in macros.

nordic-krch replied:
which one? `rc` is passed by value

pabigot commented on tests/unit/cbprintf/main.c:
```diff
@@ -371,7 +371,7 @@ static inline bool prf_check(const char *expected,
 		     NULL)
 
-static void test_pct(void)
+void test_pct(void)
```
I really don't like this commit either.  We should not have tests/examples that promote contamination of the global namespace with functions that are never referenced outside the translation unit.

If a function's commented out permanently, it should be removed.  If it's done during development, ignore the warnings.  If it's conditional on build options leave it in the list, and skip on entry when the triggering feature is disabled.

pabigot reviewed (changes requested):
Requested changes are elimination of `__C_GENERIC` and not removing `static` from internal functions.  I don't like reference macro parameters, but won't block on those.

nordic-krch committed:
lib: os: cbprintf: Add alignment check to cbprintf_package

Added validation of alignment to cbprintf_package. Error is returned if
input buffer is not aligned to the largest argument.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@pabigot comments addressed, except https://github.com/zephyrproject-rtos/zephyr/pull/32863#discussion_r591604522 which needs clarification.

nordic-krch commented:
I also added https://github.com/zephyrproject-rtos/zephyr/pull/32863/commits/ac02c989f626d5b177d2340c0206bd813f5c7c5a as a consequence of #33208 discussion. Added assert and recommended to use same alignment as for static packaging.

pabigot reviewed (approved):
Changes have been made.  I'm skeptical of the alignment stuff wonder what happens when people intersperse char arguments with long double arguments, but that isn't new so shouldn't affect this PR.

Hopefully the log rework that motivates all this will make it through review.

npitre commented:
On Wed, 10 Mar 2021, Peter A. Bigot wrote:

> I'm skeptical of the alignment stuff wonder what happens when people 
> intersperse char arguments with long double arguments, but that isn't 
> new so shouldn't affect this PR.

It's taken care of. Extra padding is inserted when needed.

Of course this is all architecture dependent so this is explicitly 
tested on all of them.
